### PR TITLE
Update to containerd v2

### PR DESCRIPTION
**Pull request by apostasie**
**Issue #, if available:**

**Description of changes:**

This PR updates to containerd v2.

**Testing performed:**

`make build` and `make test` pass locally.
Pending further testing into nerdctl.

References to cri api v1alpha could/should be removed obviously, but that might be independently of this I guess.

Lmk your toughts.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by sondavidb — COMMENTED** [review]
Hey, thanks for the contribution! We don't plan on merging until containerd v2.0.0, but this generally looks like the right direction.

Left a couple of quick questions, but otherwise looks good. Would you like me to run the CI?

File: go.mod [hunk 1]
```diff
@@ -1,11 +1,17 @@
 module github.com/awslabs/soci-snapshotter
 
-go 1.21
+go 1.22.0
+
+toolchain go1.22.4
```
[the remark this answers is not in the document]
Comment (apostasie) [reply go.mod:5]: *sigh*
Nope, just my IDE insisting on adding this...
Will remove and re-push.

File: go.mod [hunk 2]
```diff
@@ -1,11 +1,17 @@
 module github.com/awslabs/soci-snapshotter
 
-go 1.21
+go 1.22.0
```
[the remark this answers is not in the document]
Comment (apostasie) [reply go.mod:3]: Yes, correct.

I assume 1.21 will get out of support with 1.23 getting released, likely  early august, right?
That timeline might work just well - but yeah, let's see what happens with containerd.

**Comment by apostasie** [comment]
> Hey, thanks for the contribution! We don't plan on merging until containerd v2.0.0, but this generally looks like the right direction.
> 
> Left a couple of quick questions, but otherwise looks good. Would you like me to run the CI?

Thanks a lot @sondavidb 

Having the CI run would be lovely so I can fix any lingering issues with the transition.

Ok of course on merging only after ctd v2 is out - or whenever you folks feel this is ready.

Cheers.

**Comment by sondavidb** [comment]
CI looks good, we'll probably end up waiting for the 2.0 release. Thanks again!

**Comment by apostasie** [comment]
Just repushed without the extraneous `toolchain` statement.

Thanks a lot @sondavidb !

**Comment by apostasie** [comment]
@sondavidb obviously it will drift from main.
Happy to rebase whenever you need it so just ping me here

**Comment by sondavidb** [comment]
Yeah, we just had to fix some deprecated packages in #1308 to get past our linting CI, I don't foresee this happening again anytime soon so it might be best to rebase now if you don't mind @apostasie. Anything after this I won't bother you about until 2.0 comes out.

**Comment by apostasie** [comment]
Yup. Done.

Can you get the CI to go brrr?
(that type of rebase with dependencies versions flying all over always make me feel a bit nervous)

Thanks a lot @sondavidb !

**Comment by apostasie** [comment]
Yep, it is hosed.
Damn it.

**Comment by apostasie** [comment]
Repushed. Confirmed working locally. Sorry for the churn.

**Comment by sondavidb** [comment]
All good, this work is greatly appreciated on our end! Thanks again!

**Comment by apostasie** [comment]
Hey @sondavidb 

Question for you: would you be amenable to merging this PR into an ad-hoc  / dedicated branch (something like `experimental-ctdv2`)?

That way we could use this branch in nerdctl as we push through for v2 - and then we would switch the dependency back to soci main / your next release, when you folks are ready to have it on main (eg: post containerd v2 release as you said earlier).
Alternatively, we could temporarily disable soci in the first version of nerdctl v2, and re-integrate it later on, but I personally feel it would be better to get public QA / exposure for the entire feature set on nerdctl v2. 

Let me know your thoughts. 

Cheers!

**Comment by sondavidb** [comment]
I'll see what the team thinks and get back to you ASAP. As it's Friday afternoon I'm not expecting a consensus until after the weekend, though.

**Comment by apostasie** [comment]
> I'll see what the team thinks and get back to you ASAP. As it's Friday afternoon I'm not expecting a consensus until after the weekend, though.

No rush, really.
Appreciate your help! Thanks a lot and enjoy the week-end :)

**Comment by sondavidb** [comment]
Thanks, you too!

One more question — to be clear, we only need a branch, not an entire release, correct?

**Comment by apostasie** [comment]
> Thanks, you too!
> 
> One more question — to be clear, we only need a branch/commit, not an entire release, correct?

Yep, would be happy with just a branch/commit.

**Comment by sondavidb** [comment]
Hey @apostasie, apologies this slipped my mind.

Just created a new branch, [upgrade-containerd-v2](https://github.com/awslabs/soci-snapshotter/tree/upgrade-containerd-v2), you can edit the PR to be against this branch.

**Comment by apostasie** [comment]
Thanks a lot @sondavidb 
Really appreciate this!

Closing in favor of #1329

**Pull request closed without merging**